---
title: "Grey_VertTestDatabase_"
author: "Beth Davis"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

This is a cleaned version of Erin Grey's RefDB_vert_12S.R file. Original comments and documentation are preserved

# Quick Maine-eDNA Vertebrate 12S Database - test with first species 
Erin Grey, 2022-09-14, updated 2022-11-09, 2022-11-11, 2022-11-21

Notes: I (EG) created maine_vert_species-YYYY-MM-DD.csv list by merging lists from Harrison Goldspiel (New England vernal pool verts) and Renee Francolini (did not include Bacillus cereus reference, Homo sapiens, nor any of the mock species)

For this demo I (EG) just subset out a few species from each list


## Requirements:

This section describes the necessary files, file types, and set-up for the script to run properly.

* To run this database code's examples, you must have the file "maine_vert_species-YYYY-MM-DD.csv" (Erin Grey). Hereafter, examples using this file will be referred to as MaineVert examples.

* To run the code for your own files, you must have a single .csv file with the species binomial (Genus species) in a column named "species_binomial" in the header. No other columns are used. 


## Notes

* If you encounter any errors and are not sure where the problem is occurring, you can enter debug statements such as print("Check that Step 1 finished") into the code to find where the readout stops. 

* Variable names can be changed to your preferences, but make sure that all downstream mentions are also changed

* To minimize any section, use the toggle drop-down button on lefthand side. This is next to headers for text and at the very top of code chunks

* This is an RMarkdown file. To run code chunks, you can either highlight the code you wish to run and hit the 'Run' button at the top right, or use Ctrl+Enter. To run an entire code chunk, you can highlight the entire chunk and follow the previous options, or click the "Run" icon in the top right of the code chunk. For more information about RMarkdown files, see this (https://r4ds.had.co.nz/r-markdown.html)[RMarkdown Guide]

* The following step to clear global memory may be done manually, the primary purpose is to clear any lingering variables that may conflict and to ensure the location of your files matches the location of the code. When calling a file, it is recommended to use the absolute path from root. Locations where this would be useful are marked.

```{r set_space, echo = FALSE}
rm(list = ls())
```

Install and load the below packages as needed. For further documentation, most of these packages can be searched on CRAN or have documentation pulled directly in R using the "?packagename" utility.

```{r packages, include = FALSE}
if (!require("taxize")) {install.packages("taxize"); require("taxize")} # Check taxonomy
if (!require("rentrez")) {install.packages("rentrez"); require("rentrez")} # Query ENTREZ databases and download accessions
if (!require("AnnotationBustR")) {install.packages("AnnotationBustR"); require("AnnotationBustR")} # Find longest accessions, slice genes from mitogenomes
if (!require("reutils")) {install.packages("reutils"); require("reutils")}
if (!require("ape")) {install.packages("ape"); require("ape")} # Convert fasta, fastq, etc.
if (!require("ggplot2")) {install.packages("ggplot2"); require("ggplot2")} # Plots
```

The following packages rely on BiocManager/Bioconductor and installation may take a longer time or be more difficult.

```{r bio_packages, include = FALSE}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

# Install/load msa - # Multiple sequence alignment algorithms ClustalW and Muscle
BiocManager::install("msa")

# Install genbankr - parse genbank files
BiocManager::install("genbankr")

# Install BioStrings
BiocManager::install("Biostrings")

library('msa')
library('genbankr')
library('Biostrings')
```

* You must also have set up an ENTREZ API key with your GenBank account. For more information, go to [the NCBI API website](https://www.ncbi.nlm.nih.gov/home/develop/api/)

```{r api_key, echo = FALSE}
set_entrez_key(" ") # *Insert your API key in the quotation marks

# *Example:
# set_entrez_key("7c5ac035201a1835b5a81de1b74ec8613d08")
```

## Loading Variables and Files

* In this section, you will be loading your species list file as well as the primer, locus, and search term settings that future chunks will use to search GenBank for matches in taxonomy and locus.

* For locus, you can search any one (at a time) of the following:
ATP6, ATP8, COI, COII, COIII, CYTB, D_loop, ND1, ND2, ND3, ND4, ND4L, ND5, ND6, rRNA_12S, rRNA_16S, tRNA_Ala, tRNA_Arg, tRNA_Asn, tRNA_Asp, tRNA_Cys, tRNA_Gln, tRNA_Glu, tRNA_Gly, tRNA_His, tRNA_Ile, tRNA_Leu, tRNA_Lys, tRNA_Met, tRNA_Phe, tRNA_Pro, tRNA_Ser, tRNA_Thr, tRNA_Trp, tRNA_Tyr, tRNA_Val

* Refer to your own documentation or sourced papers for the sequence of your primers

If your species list contains characters outside of the English alphabet or special characters, you may encounter issues further on unless you adjust the read.csv command to:  a01_INPUT <- read.csv("CHANGE PATH", header = TRUE, encoding = 'latin1', stringsAsFactors = FALSE)

The DNAString is simply a data class used by R for genetic sequences. To see more, refer to this {https://www.rdocumentation.org/packages/Biostrings/versions/2.40.2/topics/DNAString-class}[DNAString documentation]

```{r variable_set, echo = FALSE}
a01_INPUT <- read.csv("CHANGE PATH", header = TRUE) # insert the file path and name (with .csv) of your species list file

locus = " "

primer_forward <- DNAString(" ") # Forward primer sequence
primer_reverse <- DNAString(" ") # Reverse primer sequence

ref_seq <- DNAString(" ") 
```


* Once the species file, locus, and primers are loaded, you can move to defining the search terms

* Begin by defining the reverse complements of each primer, the reference sequence, and save the length of the reference sequence

* From data(mtDNAterms) onwards, these lines use AnnotationBustR to find synonyms for the locus you chose and format them into ENTREZ-usable search terms (saved as target_locus_searchterm). If you know there are synonyms that AnnotationBustR may not have, you can also manually add them using the more_12Ssynonyms and rbind lines, with 'Name' equal to the synonym.

# QUESTION: Why do you use a reverse complement?

```{r search_set, echo = FALSE}
primer_forward_rc <- reverseComplement(primer_forward)
primer_reverse_rc <- reverseComplement(primer_reverse)
ref_seq_length <- nchar(ref_seq)
ref_seq_rc <- reverseComplement(ref_seq)


data(mtDNAterms) # This calls the premade list of synonyms from the AnnotationBustR package

# Add synonyms manually
more_12Ssynonyms <- data.frame(Locus=" ", Type=" ", Name= " ") 
mtDNAterms <- rbind(mtDNAterms, more_12Ssynonyms)

target_locus_synonyms <- mtDNAterms[mtDNAterms$Locus==locus,] 
#subset the 12S synonyms into a new data frame

target_locus_synonyms$Terms <- paste0("OR ", target_locus_synonyms$Name, "[TITL]") 
# format for ENTREZ search terms

target_locus_synonyms$Terms[1] <- paste0("AND (", target_locus_synonyms$Name[1], "[TITL]") 
# first term starts with "AND ("

target_locus_synonyms$Terms[dim(target_locus_synonyms)[1]] <- paste0("OR ", target_locus_synonyms$Name[dim(target_locus_synonyms)[1]], "[TITL])") 
#last term ends with a ")"

target_locus_searchterm <- paste(as.vector(target_locus_synonyms$Terms), collapse=" ") 
# the big 'ole ENTREZ 12S search term
```

The following code chunk is an example of the above steps using the MaineVert file.

```{r variable_set_example, echo = FALSE}
locus = "rRNA_12S" 
primer_forward <- DNAString("GTCGGTAAAACTCGTGCCAGC") # MiFish-U-F 
primer_reverse <- DNAString("CATAGTGGGGTATCTAATCCCAGTTTG") # MiFish-U-R
ref_seq <- DNAString("GCCGGTAAAACTCGTGCCAGCAACCGCGGTTAAACGAGAGGCTCTAGTTGATATACTACGGCGTAAAGGGTGGTTAAGGAACAATGTAAAATAAAGCCAAACGGCCCTTTAACTGTTATACGCTTTTAGATGCTAGAGGCCCAAACACGAAAGTAGCTTTAAATTAATTACCTGAACCCACGAAAGCTAAGAAACAAACTGGGATTAGATACCCCACTATGC") #zebrafish MiFish

a01_INPUT <- read.csv("maine_vert_species-2022-11-13.csv", header=TRUE) 

#make reverse complements, define locus search terms, etc.
primer_forward_rc <- reverseComplement(primer_forward) #reverse complement of forward primer
primer_reverse_rc <- reverseComplement(primer_reverse) #reverse complement of reverse primer
ref_seq_length <- nchar(ref_seq) # length of the reference sequence
ref_seq_rc <- reverseComplement(ref_seq) # reverse complement of reference sequence
data(mtDNAterms) #AnnotationBustR's list of synonyms for different loci
more_12Ssynonyms <- data.frame(Locus="rRNA_12S", Type="rRNA", Name= "small ribosomal RNA subunit RNA") # other synonyms that I find here
mtDNAterms <- rbind(mtDNAterms, more_12Ssynonyms)
target_locus_synonyms <- mtDNAterms[mtDNAterms$Locus==locus,] #subset the 12S synonyms into a new data frame
target_locus_synonyms$Terms <- paste0("OR ", target_locus_synonyms$Name, "[TITL]") 
target_locus_synonyms$Terms[1] <- paste0("AND (", target_locus_synonyms$Name[1], "[TITL]") 
target_locus_synonyms$Terms[dim(target_locus_synonyms)[1]] <- paste0("OR ", target_locus_synonyms$Name[dim(target_locus_synonyms)[1]], "[TITL])") #last term ends with a ")"
target_locus_searchterm <- paste(as.vector(target_locus_synonyms$Terms), collapse=" ") # the big 'ole ENTREZ 12S search term
```

## Species List Name Check

This section uses the Global Names Resolver from the Encyclopedia of Life to check the inputted a01_INPUT species list and find the best match for each species name as a quality control step against misspellings, taxonomic updates, and other issues. It outputs two .csv files - a02_NAMECHECK.csv which only only saves the best name match, and a03_BESTNAMES.csv which is a list of the best names found as well as the names from the original species list if no better match was found.


```{r name_check, echo = FALSE}
# Check animal species names using the Global Names Resolver from the Encyclopedia of Life
a02_NAMECHECK <- gnr_resolve(a01_INPUT$species_binomial, best_match_only = TRUE, canonical = TRUE, fields="all") 

# Get the best name but keep previous names (fix misspellings, use most recently accepted, etc.)
a03_BESTNAMES <- merge(a01_INPUT, a02_NAMECHECK[,c("user_supplied_name", "submitted_name","matched_name2")], by.x=c("species_binomial"), by.y=c("user_supplied_name"), all.x=TRUE) 

# Use the check & corrected name if available, if not then use the original source name
a03_BESTNAMES$search_name <- ifelse(is.na(a03_BESTNAMES$matched_name2), a03_BESTNAMES$species_binomial, a03_BESTNAMES$matched_name2) 

# Save the outputted files - Make sure the file path is set to where you want the files saved
write.csv(a02_NAMECHECK, "../Desktop/a02_NAMECHECKtest.csv", row.names=FALSE)
write.csv(a03_BESTNAMES, "../Desktop/a03_BESTNAMES_v1test.csv", row.names=FALSE)
```

# QUESTION: Does the Global Names Resolver also work for plants/non-animal species or should that be a note in the code to only use it with animal species?

## Search ENTREZ and Summarize Accession Availability

Using the user-defined search terms, this section searches the ENTREZ nucleotide database and places the accessions, mitogenomes, and target loci accession information in created columns in the a03_BESTNAMES file, printing out an updated .csv with this information.

```{r ENTREZ_prep, echo = FALSE}
# Make blank summary columns in BESTNAMES for the following:

# Column for the number of mitogenome accessions
a03_BESTNAMES$n_mitogenome <- "na" 

# Column for the number of target accessions
a03_BESTNAMES$n_target <- "na" 

# Column for the mitogenome GI numbers
a03_BESTNAMES$ids_mitogenome <- "na" 

# Column for the target accession GI numbers
a03_BESTNAMES$ids_target <- "na"

# To check that the new columns were created properly, remove the # from in front of the below line and run it
head(a03_BESTNAMES)
```

Be aware: this next step proceeds at a maximum speed of 3 seconds per row in a03_BESTNAMES in order to avoid being kicked out of the Entrez server, regardless of computer performance or specification. Allocate time properly and/or make sure your computer will not be interrupted for this task. For time estimates, calculate: ((row number * 3)/60)/60 for the number of hours

```{r ENTREZ_search, echo = FALSE}

# Search ENTREZ nucleotide database ("nucleotide"="nuccore" database)
for (i in 1:dim(a03_BESTNAMES)[1]){
  print(i) #counter
  # define search terms for species
  search_name <- paste0(a03_BESTNAMES$search_name[i],"[ORGN]") #format species name for ENTREZ search
  search_term <- paste(paste0(a03_BESTNAMES$search_name[i],"[ORGN]"), target_locus_searchterm, collapse=" ") #concatenate species and 12S search terms into one search term
  
  # Search GenBank for all accessions, mitogenomes, and target loci accessions
  mitogenomes <- entrez_search(db="nucleotide", term <- paste(search_name, "AND mitochondrion[TITL] AND complete genome[TITL]"), retmax=999999) # search for species mitogenome accessions
   Sys.sleep(1) #slow down request to the Entrez server or you'll get kicked out
  targets <- entrez_search(db="nucleotide", term <- search_term, retmax=999999) # search all species 12S accessions
   Sys.sleep(1) #slow down request to the Entrez server or you'll get kicked out

  # Update the the BESTNAMES dataframe of all accession types and associated ids (= GI numbers)
  a03_BESTNAMES$n_mitogenome[i] <-mitogenomes$count
  a03_BESTNAMES$n_target[i] <-targets$count
  a03_BESTNAMES$ids_mitogenome[i] <-paste(mitogenomes$ids, collapse="|")
  a03_BESTNAMES$ids_target[i] <- paste(targets$ids, collapse="|")
  
  # Reset the loop variables
  mitogenomes <- "na"
  targets <- "na"
  search_name <- "na"
  search_term <- "na"
}

```

```{r ENTREZ_results, echo = FALSE}

# Write out the mitogenome and target accessions for each species in an updated file
write.csv(a03_BESTNAMES, "a03_BESTNAMES_v2test.csv", row.names = FALSE) 
```


## Build Reference Database With Target Accessions

With the ENTREZ search done, you can now construct the reference database that will be used to compile and report the results. This section creates a database skeleton, adds the target accession sequences for each species in a03_BESTNAMES, and appends the results to the database (named a04_REFDB)


```{r database_construct, echo = FALSE}

# Create the database skeleton
a04_REFDB <- data.frame(seq_header=NA, sequence=NA, seq_accession=NA, type=NA, species=NA) 
```

```{r database_targetaccession, echo = FALSE}

# Add target accession sequences to database
for (j in 1:dim(a03_BESTNAMES)[1]){ #for every good species name
  print(j) #counter
  if (a03_BESTNAMES$n_target[j]>0 && a03_BESTNAMES$n_target[j]<500) { # scrape GenBank target sequences if available, but don't do Aythya affinis (Lesser Scaup) because bc server doesn't allow this
    seqs_target <- entrez_fetch(db="nuccore", id=c(unlist(strsplit(a03_BESTNAMES$ids_target[j], split="\\|"))), rettype="fasta") # fetch all the sequences from Genbank
    write(seqs_target, paste(a03_BESTNAMES$search_name[j], paste0(locus, ".fasta"))) # formatting - write out the sequences 
    fasta_target <- readDNAStringSet(paste(a03_BESTNAMES$search_name[j], paste0(locus, ".fasta")), format="fasta") #formatting - read them back in as fasta
    seqs_target_accessions <- entrez_fetch(db="nuccore", id=unlist(strsplit(a03_BESTNAMES$ids_target[j], split="\\|")), rettype="acc") # get all the 12S accession numbers
    seq_header <- names(fasta_target) #formatting
    sequence <- paste(fasta_target) #formatting
    seq_accession <- unlist(strsplit(seqs_target_accessions, split="\n")) # formatting
    tempDB <- data.frame(seq_header, sequence, seq_accession, type="accession", species=a03_BESTNAMES$search_name[j]) # make a temporary database with all sequences, their header, accession number, etc.
    
    a04_REFDB <- rbind(a04_REFDB, tempDB) # append temporary database to the full database
    
    
    # reset loop variables
    seqs_target <- "na"
    fasta_target <- "na"
    seqs_target_accessions <- "na"
    seq_header <- "na"
    sequence <- "na"
    seq_accession <- "na"
    tempDB  <- "na"
    }
} 
```

# QUESTION: Should there be a warning about this skipped species or others that GenBank can't handle?

## Add Mitogenome Scrapes to Database

This section searches for the mitogenome information for each listed species and appends that information to the database.

```{r database_mitoprep, echo = FALSE}

# Add mitogenome scrapes to database (skip Canis lupus 151, Hirundo rustica 358)
a03_BESTNAMES$n_mitogenome <- as.numeric(a03_BESTNAMES$n_mitogenome) #format
a03_BESTNAMES$n_target <- as.numeric(a03_BESTNAMES$n_target) #format

```
# QUESTION: Why were these skipped? Same question as above, should there be warnings about searching specific species in GenBank?

```{r database_mitoloop, echo = FALSE}

for (k in 1:dim(a03_BESTNAMES)[1]) { #for every good species name
  print(k) #counter
  if (a03_BESTNAMES$n_mitogenome[k]>0 && a03_BESTNAMES$n_mitogenome[k]<150) { #scrape Genbank mitogenomes if available, skip species with lots of mitogenomes bc server doesn't allow this
    mito_ids <- unlist(strsplit(a03_BESTNAMES$ids_mitogenome[k], split="\\|")) # format mitogenome ids
    mito_accessions <- entrez_fetch(mito_ids, db="nuccore", rettype="acc") # find the accession number for each mitogenome id
    mito_accessions <- unlist(strsplit(mito_accessions, split="\n")) # format accession numbers
      
      for (m in 1:length(mito_accessions)){ # loop through and scrape each mitogenome accession
        gb <- readGenBank(GBAccession(mito_accessions[m])) # get the Genbank annotation for accession 
        target_feature <- which(otherFeatures(gb)$product %in% as.character(target_locus_synonyms$Name)) # find target annotation metadata (note: use otherFeatures for for rRNAs, tRNAs, etc. and use genes(gb) for genes like COI, CYTB, NADH, etc.)
        new_row <- c(paste("Unparsed mitochondrion", mito_accessions[m], sep=" "), "na", mito_accessions[m], "scrape", species=a03_BESTNAMES$search_name[k])
        
        if(length(target_feature) > 0) { # if target feature is found in the parsed mitochondrial genome, find the sequence, otherwise say that its unparsed
          target_range <- otherFeatures(gb)@ranges[target_feature] #extract the target range information
          target_strand <-  otherFeatures(gb)@strand[target_feature] #extract the target strand information (+ or -)
          target_seq <- subseq(getSeq(gb), start=target_range@start, width=target_range@width) #scrape the genome for target
          scrapedseq_binomial <- names(target_seq) #get the binomial name
          scraped_seq <- paste(target_seq) #format
          scraped_range <- paste(target_range) #format
          new_row <- c(paste(names(target_seq),"mitochondrion", mito_accessions[m], sep=" "), paste(target_seq), mito_accessions[m], "scrape", species=a03_BESTNAMES$search_name[k])
        }
        a04_REFDB <- rbind(a04_REFDB, new_row) # update the database
    
        # reset loop variables
        rm(gb, target_feature, target_strand, target_seq, scrapedseq_binomial, scraped_seq, scraped_range, new_row)
        
        } # close m loop (each "m" accession m per species "k")
    } # close species k with mitogenomes if statement
  
  # reset loop variables
  rm(mito_ids, mito_accessions)
} # close for each species k loop


a04_REFDB <- a04_REFDB[-1,] #format - remove the top row of NAs
write.csv(a04_REFDB, "a04_REFDB.csv", row.names=FALSE)
```
# ACTION: Check if I can move the last two lines outside of this chunk. I don't like running writes over and over with loops


## Clean Reference Database of Duplicates

With the reference database compiled, it now must be cleaned and checked for duplicates. This section checks and removes duplicates in the a04_REFDB database and outputs a new database, a05_UNIQUEDB with only unique entries

```{r database_clean, echo = FALSE}

a05_UNIQUEDB <- a04_REFDB[!duplicated(a04_REFDB$sequence),] #remove duplicates

for (i in  1:dim(a05_UNIQUEDB)[1]){ # for every row in the unique db file
  dups <- subset(a04_REFDB, sequence == a05_UNIQUEDB$sequence[i]) # find identical sequences in the fill db file 
  a05_UNIQUEDB$duplicate_accessions[i] <- paste(dups$seq_accession, collapse = "|") # paste all those accessions together into a new unique db field
  a05_UNIQUEDB$duplicate_species[i] <- paste(dups[!duplicated(dups$species),"species"], collapse = "|")
}
```

```{r clean_output, echo = FALSE}
write.csv(a05_UNIQUEDB, "a05_UniqueRefDB.csv", row.names=FALSE)
```


## Align Sequences to Reference Sequence and Primers

This section takes the a05_UNIQUEDB database created above and matches the sequences in that file to primers and reference sequences, then outputs the result as a new database, a06_UniqueDB_withAlignments


```{r alignment_format}

# Format the unique database
rownames(a05_UNIQUEDB) <- 1:dim(a05_UNIQUEDB)[1]

# Format fastas
a05_UNIQUEDB$fasta <- paste(paste0(">",a05_UNIQUEDB$seq_header), a05_UNIQUEDB$sequence, sep="\n") 

# Write out the sequences
write(as.character(a05_UNIQUEDB$fasta), "db_seqs.fasta")

# Read sequences back in as fasta
target_fastas <- readDNAStringSet("db_seqs.fasta", format="fasta") 

# Create columns for different alignment variables
a05_UNIQUEDB$align_pover <- "NA"
a05_UNIQUEDB$align_pid <- "NA"
a05_UNIQUEDB$align_pover_rc <- "NA"
a05_UNIQUEDB$align_pid_rc <- "NA"
a05_UNIQUEDB$nmismatch_forward <- "NA"
a05_UNIQUEDB$nmismatch_forward_clamp <- "NA"
a05_UNIQUEDB$nmismatch_reverse_rc <- "NA"
a05_UNIQUEDB$nmismatch_reverse_clamp_rc <- "NA"
a05_UNIQUEDB$nmismatch_forward_rc <- "NA"
a05_UNIQUEDB$nmismatch_forward_clamp_rc <- "NA"
a05_UNIQUEDB$nmismatch_reverse <- "NA"
a05_UNIQUEDB$nmismatch_reverse_clamp <- "NA"
```

```{r alignment_loop, echo = FALSE}
for (i in 1:dim(a05_UNIQUEDB)[1]){ 
  print(i)
  ## align to reference sequence, calculate basic stats, and save to a05_UNIQUEDBaframe
  temp_align <- pairwiseAlignment(ref_seq, target_fastas[i], gapOpening = 10, gapExtension = 4, type="global", scoreOnly=FALSE) #align with pairwiseAlignment()
  a05_UNIQUEDB$align_pover[i] <- nchar(temp_align)/ref_seq_length # calculate percent of reference sequence aligned to the target
  a05_UNIQUEDB$align_pid[i] <- pid(temp_align)  # calculate percent identity of aligned reference sequence to the target
  temp_align <- "NA" #reset the variable
  
  ## align as above, but to the reverse complement
  temp_align_rc <- pairwiseAlignment(ref_seq_rc, target_fastas[i],  gapOpening = 10, gapExtension = 4, type="global", scoreOnly=FALSE) #align with pairwiseAlignment()
  a05_UNIQUEDB$align_pover_rc[i] <- nchar(temp_align_rc)/ref_seq_length # calculate percent of reference sequence aligned to the target
  a05_UNIQUEDB$align_pid_rc[i] <- pid(temp_align_rc)# calculate percent identity of aligned reference sequence to the target
  temp_align_rc <- "NA" #reset the variable
  
  ## primer matching
  test_string <- DNAString(as.character(target_fastas[i])) #matchPattern needs it in this format for some reason
  #forward primer match
  match_forward <- matchPattern(primer_forward, test_string, max.mismatch=6, with.indels=FALSE) #match the primer to the sequence
  if (length(match_forward) == 1){ # if there is a match, fill in the following info
    mismatch_forward <- do.call(rbind, mismatch(primer_forward, match_forward)) #locations of mismatches on the primer
    a05_UNIQUEDB$nmismatch_forward[i] <- nmismatch(primer_forward, match_forward) #number of mismatches to the primer
    a05_UNIQUEDB$nmismatch_forward_clamp[i] <- length(which(mismatch_forward > (length(primer_forward)-6))) #number of mismatches in gc_clamp
  }
  if (length(match_forward) > 1){
    mismatch_forward <- do.call(rbind, mismatch(primer_forward, match_forward)) #locations of mismatches on the primer
    temp_id <- which.min(rowSums(mismatch_forward < 6)) # find the index of the match with the smallest number of mismatches in the clamp
    a05_UNIQUEDB$nmismatch_forward[i] <- nmismatch(primer_forward, match_forward[temp_id]) #number of mismatches to the primer
    a05_UNIQUEDB$nmismatch_forward_clamp[i] <- length(which(mismatch_forward[temp_id,] < 6))
    rm(temp_id)
  }
  
  #reverse complement of forward primer
  match_forward_rc <- matchPattern(primer_forward_rc, test_string, max.mismatch=6, with.indels=FALSE) #match the primer to the sequence
  if (length(match_forward_rc) == 1){
    mismatch_forward_rc <- do.call(rbind, mismatch(primer_forward_rc, match_forward_rc)) #locations of mismatches on the primer
    a05_UNIQUEDB$nmismatch_forward_rc[i] <- nmismatch(primer_forward_rc, match_forward_rc) #number of mismatches to the primer
    a05_UNIQUEDB$nmismatch_forward_clamp_rc[i] <- length(which(mismatch_forward_rc < 6))
  }
  if (length(match_forward_rc) > 1){
    mismatch_forward_rc <- do.call(rbind, mismatch(primer_forward_rc, match_forward_rc)) #locations of mismatches on the primer
    temp_id <- which.min(rowSums(mismatch_forward_rc < 6)) # find the index of the match with the smallest number of mismatches in the clamp
    a05_UNIQUEDB$nmismatch_forward_rc[i] <- nmismatch(primer_forward_rc, match_forward_rc[temp_id]) #number of mismatches to the primer
    a05_UNIQUEDB$nmismatch_forward_clamp_rc[i] <- length(which(mismatch_forward_rc[temp_id,] < 6))
    rm(temp_id)
  }
  
  #reverse primer
  match_reverse <- matchPattern(primer_reverse, test_string, max.mismatch=6, with.indels=FALSE) #match the primer to the sequence
  if (length(match_reverse) == 1){
    mismatch_reverse <- do.call(rbind, mismatch(primer_reverse, match_reverse)) #locations of mismatches on the primer
    a05_UNIQUEDB$nmismatch_reverse[i] <- nmismatch(primer_reverse, match_reverse) #number of mismatches to the primer
    a05_UNIQUEDB$nmismatch_reverse_clamp[i] <- length(which(mismatch_reverse > (length(primer_reverse)-6))) #number of mismatches in gc_clamp
  }
  if (length(match_reverse) > 1){
    mismatch_reverse <- do.call(rbind, mismatch(primer_reverse, match_reverse)) #locations of mismatches on the primer
    temp_id <- which.min(rowSums(mismatch_reverse < 6)) # find the index of the match with the smallest number of mismatches in the clamp
    a05_UNIQUEDB$nmismatch_reverse[i] <- nmismatch(primer_reverse, match_reverse[temp_id]) #number of mismatches to the primer
    a05_UNIQUEDB$nmismatch_reverse_clamp[i] <- length(which(mismatch_reverse[temp_id,] < 6))
    rm(temp_id)
  }
  #reverse complement of reverse primer
  match_reverse_rc <- matchPattern(primer_reverse_rc, test_string, max.mismatch=6, with.indels=FALSE) #match the primer to the sequence
  if (length(match_reverse_rc) == 1){
    mismatch_reverse_rc <- do.call(rbind, mismatch(primer_reverse_rc, match_reverse_rc)) #locations of mismatches on the primer
    a05_UNIQUEDB$nmismatch_reverse_rc[i] <- nmismatch(primer_reverse_rc, match_reverse_rc) #number of mismatches to the primer
    a05_UNIQUEDB$nmismatch_reverse_clamp_rc[i] <- length(which(mismatch_reverse_rc < 6))
  }
  if (length(match_reverse_rc) > 1){
    mismatch_reverse_rc <- do.call(rbind, mismatch(primer_reverse_rc, match_reverse_rc)) #locations of mismatches on the primer
    temp_id <- which.min(rowSums(mismatch_reverse_rc < 6)) # find the index of the match with the smallest number of mismatches in the clamp
    a05_UNIQUEDB$nmismatch_reverse_rc[i] <- nmismatch(primer_reverse_rc, match_reverse_rc[temp_id]) #number of mismatches to the primer
    a05_UNIQUEDB$nmismatch_reverse_clamp_rc[i] <- length(which(mismatch_reverse_rc[temp_id,] < 6))
    rm(temp_id)
  }
  
  test_string <-"NA"
  match_forward <- "NA"
  mismatch_forward <- "NA"
  match_forward_rc <- "NA"
  mismatch_forward_rc <- "NA"
  match_reverse <- "NA"
  mismatch_reverse <- "NA"
  match_reverse_rc <- "NA"
  mismatch_reverse_rc <- "NA"
}

a06_UniqueDB_withAlignments <- a05_UNIQUEDB
write.csv(a06_UniqueDB_withAlignments, "a06_UniqueDB_withAlignments.csv", row.names = FALSE)
```


## Update Summary Dataframe

# ACTION: Ask Erin what this is doing exactly, or fill it in after I test it

```{r format, echo = FALSE}
## format primer alignment fields
a06_UniqueDB_withAlignments$nmismatch_forward <- as.numeric(a06_UniqueDB_withAlignments$nmismatch_forward) # format
a06_UniqueDB_withAlignments$nmismatch_forward_clamp <- as.numeric(a06_UniqueDB_withAlignments$nmismatch_forward_clamp)# format
a06_UniqueDB_withAlignments$nmismatch_reverse_rc <- as.numeric(a06_UniqueDB_withAlignments$nmismatch_reverse_rc)# format
a06_UniqueDB_withAlignments$nmismatch_reverse_clamp_rc <- as.numeric(a06_UniqueDB_withAlignments$nmismatch_reverse_clamp_rc)# format
a06_UniqueDB_withAlignments$nmismatch_forward_rc <- as.numeric(a06_UniqueDB_withAlignments$nmismatch_forward_rc)# format
a06_UniqueDB_withAlignments$nmismatch_forward_clamp_rc <- as.numeric(a06_UniqueDB_withAlignments$nmismatch_forward_clamp_rc)# format
a06_UniqueDB_withAlignments$nmismatch_reverse <- as.numeric(a06_UniqueDB_withAlignments$nmismatch_reverse)# format
a06_UniqueDB_withAlignments$nmismatch_reverse_clamp <- as.numeric(a06_UniqueDB_withAlignments$nmismatch_reverse_clamp)# format
```

```{r count_mismatch, echo = FALSE}
## count total mismatches for both potential primer pairs & then choose the primer pair with fewest mismatches
a06_UniqueDB_withAlignments$nmismatch_pair_forward <- a06_UniqueDB_withAlignments$nmismatch_forward + a06_UniqueDB_withAlignments$nmismatch_reverse_rc # count mismatches in forward/reverse_rc
a06_UniqueDB_withAlignments$nmismatch_pair_forward_rc <- a06_UniqueDB_withAlignments$nmismatch_forward_rc + a06_UniqueDB_withAlignments$nmismatch_reverse #count mismatches in forward_rc/reverse
```

```{r primer_stuff, echo = FALSE}
a06_UniqueDB_withAlignments$primer_pair <- "NA" #figure out which primer pair to choose (forward, forward_rc, or tie) with following decision tree, if no primer match keep as "NA"
for (x in 1:dim(a06_UniqueDB_withAlignments)[1]) {
  if (!is.na(a06_UniqueDB_withAlignments$nmismatch_pair_forward[x]) &&  is.na(a06_UniqueDB_withAlignments$nmismatch_pair_forward_rc[x])) { #if forward primer match exists and forward_rc match doesn't, use forward
   a06_UniqueDB_withAlignments$primer_pair[x] <- "forward" #use the forward primer
  } else if (is.na(a06_UniqueDB_withAlignments$nmismatch_pair_forward[x]) && !is.na(a06_UniqueDB_withAlignments$nmismatch_pair_forward_rc[x])) { # forward [rimer match doesn't exist and forward_rc match does, use forward_rc
    a06_UniqueDB_withAlignments$primer_pair[x] <- "forward_rc"  #use the forward_rc primer
  } else if (is.na(a06_UniqueDB_withAlignments$nmismatch_pair_forward[x]) &&  is.na(a06_UniqueDB_withAlignments$nmismatch_pair_forward_rc[x])) { # neither primer matches, use none
    a06_UniqueDB_withAlignments$primer_pair[x] <- "na" 
  } else if (!is.na(a06_UniqueDB_withAlignments$nmismatch_pair_forward[x]) && !is.na(a06_UniqueDB_withAlignments$nmismatch_pair_forward_rc[x]) && a06_UniqueDB_withAlignments$nmismatch_pair_forward[x] < a06_UniqueDB_withAlignments$nmismatch_pair_forward_rc[x]) { # if both matches exist, choose forward if less mismatches
    a06_UniqueDB_withAlignments$primer_pair <- "forward" 
  } else if (!is.na(a06_UniqueDB_withAlignments$nmismatch_pair_forward[x]) && !is.na(a06_UniqueDB_withAlignments$nmismatch_pair_forward_rc[x]) && a06_UniqueDB_withAlignments$nmismatch_pair_forward_rc[x] < a06_UniqueDB_withAlignments$nmismatch_pair_forward[x]) { # if both matches exist, choose forward_rc if less mismatches
    a06_UniqueDB_withAlignments$primer_pair[x] <- "forward_rc" 
  } else if (!is.na(a06_UniqueDB_withAlignments$nmismatch_pair_forward[x]) && !is.na(a06_UniqueDB_withAlignments$nmismatch_pair_forward_rc[x]) && a06_UniqueDB_withAlignments$nmismatch_pair_forward[x] == a06_UniqueDB_withAlignments$nmismatch_pair_forward_rc[x]) { # if both matches exist and equal mismatches,
    a06_UniqueDB_withAlignments$primer_pair <- "tie" 
  }
}

a06_UniqueDB_withAlignments$primer_pair <- as.factor(a06_UniqueDB_withAlignments$primer_pair) # make primer_pair field a factor

a06_UniqueDB_withAlignments$align_pover_final <- "NA"
a06_UniqueDB_withAlignments$align_pid_final <- "NA"
a06_UniqueDB_withAlignments$nmismatch_forward_final <- "NA"
a06_UniqueDB_withAlignments$nmismatch_forward_clamp_final <- "NA"
a06_UniqueDB_withAlignments$nmismatch_reverse_final <- "NA"
a06_UniqueDB_withAlignments$nmismatch_reverse_clamp_final <- "NA"

for (w in 1:dim(a06_UniqueDB_withAlignments)[1]){ # if best primer pair is forward or a tie, update with primer match info (need to state that will go with forward pair if tie somewhere)
  if (a06_UniqueDB_withAlignments$primer_pair[w] == "forward" || a06_UniqueDB_withAlignments$primer_pair[w] == "tie"){
    a06_UniqueDB_withAlignments$align_pover_final[w] <- a06_UniqueDB_withAlignments$align_pover[w]
    a06_UniqueDB_withAlignments$align_pid_final[w] <- a06_UniqueDB_withAlignments$align_pid[w]
    a06_UniqueDB_withAlignments$nmismatch_forward_final[w] <- a06_UniqueDB_withAlignments$nmismatch_forward[w]
    a06_UniqueDB_withAlignments$nmismatch_forward_clamp_final[w] <- a06_UniqueDB_withAlignments$nmismatch_forward_clamp[w]
    a06_UniqueDB_withAlignments$nmismatch_reverse_final[w] <- a06_UniqueDB_withAlignments$nmismatch_reverse_rc[w]
    a06_UniqueDB_withAlignments$nmismatch_reverse_clamp_final[w] <- a06_UniqueDB_withAlignments$nmismatch_reverse_clamp_rc[w]
  } else if (a06_UniqueDB_withAlignments$primer_pair[w] == "forward_rc"){ # if best primer pair is forward_rc, update withh primer match info
    a06_UniqueDB_withAlignments$align_pover_final[w] <- a06_UniqueDB_withAlignments$align_pover_rc[w]
    a06_UniqueDB_withAlignments$align_pid_final[w] <- a06_UniqueDB_withAlignments$align_pid_rc[w]
    a06_UniqueDB_withAlignments$nmismatch_forward_final[w] <- a06_UniqueDB_withAlignments$nmismatch_forward_rc[w]
    a06_UniqueDB_withAlignments$nmismatch_forward_clamp_final[w] <- a06_UniqueDB_withAlignments$nmismatch_forward_clamp_rc[w]
    a06_UniqueDB_withAlignments$nmismatch_reverse_final[w] <- a06_UniqueDB_withAlignments$nmismatch_reverse[w]
    a06_UniqueDB_withAlignments$nmismatch_reverse_clamp_final[w] <- a06_UniqueDB_withAlignments$nmismatch_reverse_clamp[w]
  }
}


#Count reference sequences by species
counts_species <- as.data.frame(table(a06_UniqueDB_withAlignments$species)) #count the number of reference sequences per species
colnames(counts_species) <- c("species", "unique_target_seqs_n") #format
a03_BESTNAMES <- merge(a03_BESTNAMES, counts_species, by.x= "search_name", by.y="species", all.x=TRUE, all.y=TRUE) # update the summary database with reference sequence counts

#Count reference sequences that aligning to primer with <6 mismatches by species
matched_refs <- (a06_UniqueDB_withAlignments[a06_UniqueDB_withAlignments$primer_pair != "na",]) # sequencs with a primer pair match
counts_species_matched <- as.data.frame(table(matched_refs$species)) # count number of matched sequences by species
colnames(counts_species_matched) <- c("species", "primer_matched_n") #format
a03_BESTNAMES <- merge(a03_BESTNAMES, counts_species_matched, by.x= "search_name", by.y="species", all.x=TRUE, all.y=TRUE) # add count to summary database

#Reference sequences aligning to primer with <6 mismatches and additional threshold(s)
threshold1 <- 1
matched_refs_threshold1 <- matched_refs[(matched_refs$nmismatch_forward_clamp_final <= threshold1) && (matched_refs$nmismatch_reverse_clamp_final <= threshold1),]
counts_species_threshold1 <- as.data.frame(table(matched_refs_threshold1$species))
colnames(counts_species_threshold1) <- c("species", "clamp_threshold1_n") #format
a03_BESTNAMES <- merge(a03_BESTNAMES, counts_species_threshold1, by.x= "search_name", by.y="species", all.x=TRUE, all.y=TRUE)

threshold2 <- 2
matched_refs_threshold2 <- matched_refs[(matched_refs$nmismatch_forward_clamp_final <= threshold2) && (matched_refs$nmismatch_reverse_clamp_final <= threshold2),]
counts_species_threshold2 <- as.data.frame(table(matched_refs_threshold2$species))
colnames(counts_species_threshold2) <- c("species", "clamp_threshold2_n") #format
a03_BESTNAMES <- merge(a03_BESTNAMES,counts_species_threshold2, by.x= "search_name", by.y="species", all.x=TRUE, all.y=TRUE)

#add in duplicates information to best_names
a03_BESTNAMES$dup_accessions <- "na"
a03_BESTNAMES$dup_species <- "na"
a03_BESTNAMES$dup_species_n <- "na"
for (b in 1:dim(a03_BESTNAMES)[1]){
  dup_accessions <- paste(a06_UniqueDB_withAlignments[which(a06_UniqueDB_withAlignments$species == a03_BESTNAMES$search_name[b]), "duplicate_accessions"], collapse="|")
  dup_accessions <- unlist(strsplit(dup_accessions, split="\\|"))
  dup_accessions_unique <- paste(unique(dup_accessions), collapse="|")
  
  dup_species <- paste(a06_UniqueDB_withAlignments[which(a06_UniqueDB_withAlignments$species == a03_BESTNAMES$search_name[b]), "duplicate_species"], collapse="|")
  dup_species <- unlist(strsplit(dup_species, split="\\|"))
  dup_species_unique <- unique(dup_species)
  
    if (length(dup_accessions)>0){
  a03_BESTNAMES$dup_accessions[b]<-dup_accessions_unique
  a03_BESTNAMES$dup_species[b]<- paste(dup_species_unique, collapse="|")
  a03_BESTNAMES$dup_species_n[b]<- length(dup_species_unique)
  }
  
  #reset variables
  rm(dup_accessions)
  rm(dup_species)
}

a03_BESTNAMES$n_target_all <- a03_BESTNAMES$n_target + a03_BESTNAMES$n_mitogenome
write.csv(a03_BESTNAMES, "a03_BESTNAMES_v3.csv", row.names=FALSE)
```


## Summarize Results

This section summarizes the output results - total species, species with target sequences, primer matches, and/or clamp thresholds. Congrats you're done!

```{r stage_results, echo = FALSE}
n_species <- dim(a03_BESTNAMES)[1]
n_species_target <- length(which(a03_BESTNAMES$n_target_all>0))
n_species_primermatch <- length(which(a03_BESTNAMES$primer_matched_n>0)) 
n_species_clamp1 <- length(which(a03_BESTNAMES$clamp_threshold1_n>0))
n_species_clamp2 <- length(which(a03_BESTNAMES$clamp_threshold2_n>0)) 
```

```{r summary, echo = FALSE}
n_species; n_species_target; n_species_primermatch; n_species_clamp1; n_species_clamp2
```
```
